File: futbot-test/tracker.py
import cv2
import numpy as np
from config import TRACKER_TYPE, FRAME_WIDTH, FRAME_HEIGHT


_TRACKER_AVAILABLE: bool | None = None  # lazily detected


def _make_tracker():
    """Returns a tracker instance or None if opencv-contrib is not installed."""
    global _TRACKER_AVAILABLE
    try:
        if TRACKER_TYPE == "MOSSE":
            t = cv2.legacy.TrackerMOSSE_create()
        else:
            t = cv2.legacy.TrackerKCF_create()
        if _TRACKER_AVAILABLE is None:
            _TRACKER_AVAILABLE = True
        return t
    except AttributeError:
        if _TRACKER_AVAILABLE is None:
            _TRACKER_AVAILABLE = False
            print("[tracker] cv2.legacy not available (opencv-contrib required) — tracker disabled, using Kalman only")
        return None


class BallTracker:
    """
    Wraps OpenCV tracker. Used between AI inference frames.
    Call init() when we have a confident detection.
    Call update() on every frame for fast position updates.
    Falls back to no-op if opencv-contrib is not installed.
    """

    def __init__(self):
        self._tracker = None
        self._active = False

    def init(self, frame: np.ndarray, cx: int, cy: int, radius: int):
        t = _make_tracker()
        if t is None:
            return
        r = max(radius, 10)
        x = max(0, cx - r)
        y = max(0, cy - r)
        w = min(2 * r, FRAME_WIDTH - x)
        h = min(2 * r, FRAME_HEIGHT - y)
        self._tracker = t
        self._tracker.init(frame, (x, y, w, h))
        self._active = True

    def update(self, frame: np.ndarray) -> tuple[int, int] | None:
        """Returns (cx, cy) or None if tracking lost."""
        if not self._active or self._tracker is None:
            return None
        ok, bbox = self._tracker.update(frame)
        if not ok:
            self._active = False
            return None
        x, y, w, h = [int(v) for v in bbox]
        return x + w // 2, y + h // 2

    def reset(self):
        self._active = False
        self._tracker = None

File: App/futbot-test-rs-kernel-v2/src/tracker.rs
//! OpenCV MOSSE/KCF tracker — mirrors tracker.py exactly.
//!
//! Wraps an OpenCV legacy tracker for fast inter-frame updates between
//! AI inference frames. Degrades gracefully if opencv-contrib is absent
//! (build without the `opencv-tracking` cargo feature).

use opencv::core::Mat;

#[cfg(feature = "opencv-tracking")]
use crate::config::{FRAME_HEIGHT, FRAME_WIDTH};

pub struct BallTracker {
    #[cfg(feature = "opencv-tracking")]
    inner: Option<TrackerInner>,
    #[cfg(not(feature = "opencv-tracking"))]
    _phantom: (),
}

#[cfg(feature = "opencv-tracking")]
struct TrackerInner {
    tracker: Box<dyn opencv::tracking::TrackerTrait>,
    active: bool,
}

impl BallTracker {
    pub fn new() -> Self {
        #[cfg(feature = "opencv-tracking")]
        return BallTracker { inner: None };

        #[cfg(not(feature = "opencv-tracking"))]
        {
            log::debug!("[tracker] opencv-tracking feature not enabled — tracker disabled, using Kalman only");
            BallTracker { _phantom: () }
        }
    }

    /// Initialize tracker with ball bounding box. Mirrors `init()` in Python.
    pub fn init(&mut self, frame: &Mat, cx: i32, cy: i32, radius: i32) {
        #[cfg(feature = "opencv-tracking")]
        {
            use crate::config::TRACKER_TYPE;
            use opencv::core::Rect2d;

            let result = if TRACKER_TYPE == "MOSSE" {
                opencv::tracking::legacy::TrackerMOSSE::create()
                    .map(|t| Box::new(t) as Box<dyn opencv::tracking::TrackerTrait>)
            } else {
                opencv::tracking::legacy::TrackerKCF::create()
                    .map(|t| Box::new(t) as Box<dyn opencv::tracking::TrackerTrait>)
            };

            let mut tracker = match result {
                Ok(t) => t,
                Err(e) => {
                    log::warn!("[tracker] create failed: {} — tracker disabled", e);
                    return;
                }
            };

            let r = radius.max(10) as f64;
            let x = (cx as f64 - r).max(0.0);
            let y = (cy as f64 - r).max(0.0);
            let w = (2.0 * r).min(FRAME_WIDTH as f64 - x);
            let h = (2.0 * r).min(FRAME_HEIGHT as f64 - y);

            match tracker.init(frame, Rect2d::new(x, y, w, h)) {
                Ok(_) => {
                    self.inner = Some(TrackerInner {
                        tracker,
                        active: true,
                    });
                }
                Err(e) => {
                    log::warn!("[tracker] init failed: {}", e);
                }
            }
        }

        #[cfg(not(feature = "opencv-tracking"))]
        {
            let _ = (frame, cx, cy, radius);
        }
    }

    /// Update tracker on new frame. Returns center (cx, cy) or None if lost.
    pub fn update(&mut self, frame: &Mat) -> Option<(i32, i32)> {
        #[cfg(feature = "opencv-tracking")]
        {
            use opencv::core::Rect2d;
            let inner = self.inner.as_mut()?;
            if !inner.active {
                return None;
            }
            let mut bbox = Rect2d::default();
            match inner.tracker.update(frame, &mut bbox) {
                Ok(true) => {
                    let cx = (bbox.x + bbox.width / 2.0) as i32;
                    let cy = (bbox.y + bbox.height / 2.0) as i32;
                    Some((cx, cy))
                }
                _ => {
                    inner.active = false;
                    None
                }
            }
        }

        #[cfg(not(feature = "opencv-tracking"))]
        {
            let _ = frame;
            None
        }
    }

    pub fn reset(&mut self) {
        #[cfg(feature = "opencv-tracking")]
        {
            self.inner = None;
        }
    }
}

impl Default for BallTracker {
    fn default() -> Self {
        Self::new()
    }
}
